Workspaces of Sourcegraph blogpost (#2527)
sourcegraph/about: blogposts/2021/workspaces-of-sourcegraph.md

---
title: The workspaces of Sourcegraph
author: Rob Rhyne
authorUrl: https://twitter.com/robrhyne
publishDate: 2021-02-10T18:00+02:00
tags: [blog, remote]
slug: workspaces-of-sourcegraph
published: true
---

> This started as a design research thread in Slack and escalated into something awesome very quickly! Here are all the workspaces and dev setups of our team.

Reviewer: @rrhyne can you add some context to the introduction? It would be interesting to know what you are researching that prompted the question and/or why a great workspace is so important.

![Workspace 1 of Noah Santschi-Cooney, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/noah-sc-1.png)
![Workspace 2 of Noah Santschi-Cooney, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/noah-sc-2.png)

## Eric Fritz, Software Engineer

- VSCode with [Fira Code](https://github.com/tonsky/FiraCode) and [Nord Extra Dark](https://marketplace.visualstudio.com/items?itemName=yamenarahman.nord-extra-dark), [Rainbow Brackets](https://marketplace.visualstudio.com/items?itemName=2gua.rainbow-brackets), and [Indent-Rainbow](https://marketplace.visualstudio.com/items?itemName=oderwat.indent-rainbow)
- Dark mode
- 27" 5k iMac (3.2 GHz 8-core Intel Xeon W, 64GB RAM) and 2x Dell Ultrasharp U2718Q 27" 4K IPS monitors Portrait/Landscape/Portrait mode
- Ergodox EZ with Cherry MX Blue switches
- My [dotfiles](https://github.com/efritz/dotfiles)

![Combined workspace of Eric Fritz, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/efritz-combined.png)

## Adam Harvey, Software Engineer

> I use a mixture of VS Code with more or less the recommended plugins and nothing else (sometimes in remote mode to a devbox in the cloud; sometimes locally), and neovim with airline (for a better statusbar) and coc (for LSP purposes) — I've tried the native support in neovim nightlies, but I hit too many rough edges. Everything is in light mode. Like Keegan, I'm also very keyboard driven.
> 
> I don't tend to layer much else into my editing experience: I use a separate terminal (kitty), rather than the builtin ones.
> 
> Outside of that, I tend to be very single window focused: I run everything in a separate fullscreen window on a separate workspace in Sway, and use hotkeys to switch between workspaces. (1 is always Firefox; 2 is kitty; 3 is personal chat; 4 is Slack; 5 is VSCode; 6- are used for ad hoc one offs.) Generally speaking I have as little chrome as possible: menu bars are usually disabled once I know hot keys, and I have sway configured to disable titlebars unless the workspace is split or the window is floating. Minimal distraction.
> 
> ![Workspace 1 of Adam Harvey, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/aharvey.png)

Reviewer 2: Cleaning up wording, capitalisation, and linking things (debating whether to put a "btw i use arch" joke in):

```suggestion
I use a mixture of [Visual Studio Code](https://code.visualstudio.com/) with more or less the recommended language support plugins and nothing else (sometimes in remote mode to a devbox in the cloud; sometimes locally), and [Neovim](https://neovim.io/) with [airline](https://github.com/vim-airline/vim-airline) for a better statusbar and [coc.nvim](https://github.com/neoclide/coc.nvim) to get the same LSP magic as you get in VSCode — I've tried the [native LSP client](https://neovim.io/doc/user/lsp.html) in neovim nightlies, but I hit too many rough edges. I tend to keep my editors as just editors: I don't use builtin terminals or debugging capabilities.

Styling wise, everything is in light mode — I find it helps me focus, and I keep dark mode for late night phone usage when I should be sleeping — and I use [Fira Code](https://github.com/tonsky/FiraCode) as a font because I love ligatures, even if I don't always love its serifs. Like Keegan, I'm also very keyboard driven.

In terms of general workflow, I tend to be very single window focused: I run everything in a separate fullscreen window on a separate workspace in [Sway](https://swaywm.org/) on [Arch Linux](https://archlinux.org/), and use hotkeys to switch between workspaces. (1 is always Firefox; 2 is kitty; 3 is personal chat; 4 is Slack; 5 is VSCode; 6- are used for ad hoc one offs.) Generally speaking I have as little chrome as possible, besides [Waybar](https://github.com/Alexays/Waybar): menu bars are usually disabled once I know hot keys, and I have Sway configured to disable title bars unless the workspace has multiple tiled windows or the window is floating. Minimal distraction.

![vim workspace of Adam Harvey, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/aharvey-1.png)
![VSCode workspace of Adam Harvey, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/aharvey-2.png)
```

I'm also going to put a better screenshot in.

- Console: zsh + Starship Theme + fzf + bat = :heart:

![Workspace of Gonzalo Peci, Engineering Manager](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/gonza.png)

## Thorsten Ball, Software Engineer

- Environment: tmux in a fullscreen Kitty window that covers everything
- Editor: Neovim (sometimes VS Code with Vim keybindings for TypeScript/Rust)
- Mode: 90% of the time it’s light with the Lucius colorscheme, 10% it’s dark with the VS Code dark colorscheme
- Font: JetBrains Mono
- Shell: zsh
- Vim configuration: https://github.com/mrnugget/vimconfig
- Dotfiles: https://github.com/mrnugget/dotfiles

I basically do everything either in tmux or in the browser. In tmux I spawn lots of windows/shells as I need them, but have different sessions for work, private recreational programming, and other things.  In the browser I always have 3 pinned tabs (for gmail accounts and other email account). I also use Alfred a lot (to search Sourcegraph for example).

![Workspace of Thorsten Ball, Engineering Manager](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/thorsten.png)

> ## Thorsten Ball, Software Engineer

Reviewer 3: Two Thorsten?

Reviewer 4: Fixed, thanks!

- Environment: tmux in a fullscreen Kitty window that covers everything
- Editor: Neovim (sometimes VS Code with Vim keybindings for TypeScript/Rust)
- Mode: 90% of the time it’s light with the Lucius colorscheme, 10% it’s dark with the VS Code dark colorscheme
- Font: JetBrains Mono
- Shell: zsh
- Vim configuration: https://github.com/mrnugget/vimconfig
- Dotfiles: https://github.com/mrnugget/dotfiles

I basically do everything either in tmux or in the browser. In tmux I spawn lots of windows/shells as I need them, but have different sessions for work, private recreational programming, and other things.  In the browser I always have 3 pinned tabs (for gmail accounts and other email account). I also use Alfred a lot (to search Sourcegraph for example).

![Workspace of Thorsten Ball, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/thorsten.png)

> ## TJ Kandala, Software Engineer
> 
> - Editor: VS Code
> - Mode: Dark
> - Theme: SynthWave 84
> - Extensions: Bracket Pair Colorizer, GitLens, Sourcegraph
> - Misc: Fira Code font, Tilix terminal emulator
> 
> ![Workspace of Thorsten Ball, Engineering Manager](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/tj.png)

Reviewer 3: ```suggestion
## TJ Kandala, Software Engineer

- Editor: VS Code
- Mode: Dark
- Theme: SynthWave 84
- Extensions: Bracket Pair Colorizer, GitLens, Sourcegraph
- Misc: Fira Code font, Tilix terminal emulator

![Workspace of TJ Kandala, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/tj.png)
```

---

> title: The workspaces of Sourcegraph

Reviewer: We should consider a broader title (I borrowed this title to intro the workspaces section). From my perspective, the essence of the post is about how developers prefer to customize/personalize their workspaces. Here are some ideas:

- What we found after analyzing X# of developer workspaces 
- Why developers prefer to personalize their dev workspaces (needs to answer this question in detail in the intro) 
- How to improve the display of code (needs to answer this question in detail - may be a follow up post) 
- Analyzing the workspaces of Sourcegraph 
- What a developer's workspace says about their workflow 
- What makes developer workspaces unique?

Reviewer 5: When I read this title I thought it was about home office setups

author: Rob Rhyne
authorUrl: https://twitter.com/robrhyne
publishDate: 2021-02-10T18:00+02:00
tags: [blog, remote, design]
slug: workspaces-of-sourcegraph
published: true
---

> The Sourcegraph design team is exploring how to improve the display of code throughout the application. As part of this exploration, we are reviewing trends in typography, color schemes, and UI complexity in the places our engineers view code the most… their editors.  While our research is far from complete, we thought you might find a breakdown of the visual characteristics of where our team writes code of interest. 

Reviewer: why is improving the display of code throughout the application a priority for you? What problem are you addressing and do you expect the outcome of focusing on this issue to be? The reason I ask and think we can expand here is that I think the research you are doing has broad appeal. Right now, it's still pretty centered on Sourcegraph. Think of Sourcegraph as a character in your story but not the whole plot. This will make the article more interesting to more people.

## Light/dark mode and contrast

> Three quarters of our engineers preferred a dark theme over a light theme.  Curiously, the majority of dark mode users chose a low contrast between code and background. Backgrounds colors averaged 85% black while the  primary code color to background contrast ratio was around 9 to 1. 

Reviewer: This is a great data point! Can you lead into it? What is interesting about studying light/dark mode and contrast? Why is it "curious" that dark mode users chose a low contrast?  The answer may seem intuitive or obvious to you but including a sentence explaining will make sure the % of readers who don't intuitively know this is interesting stay with you. Keeping it to about a sentence will make sure readers who do share your intuition are not bored.

Conversely, all light mode users at Sourcegraph prefer the maximum contrast with 100% white backgrounds and a black primary code color. 

![Representative contrast](https://sourcegraphstatic.com/blog/workspaces-of-sourcegraph/workspace-contrast.png)
*Contrast representative of typical dark and light modes.*

## Information density

> There are clearly two camps… those who use an IDE (mainly VS Code) and terminal based editors such as neovim/vim/emacs users. Editor users had around 96% of editor screen real estate dedicated to code. IDE users had from 80 to 90% of their screens is dedicated to code or the terminal, with the rest occupied by file trees and other UI such as menus and tabs. 

Reviewer: Spell out IDE on the first reference. Again, we want to make sure that readers who may be less educated on the topic can follow along. 

Is there any insight you can share of why people prefer one or the other? Does that help us understand the real estate question? Does using one or the other say anything about a person's workflow?

## Fonts are commonly customized

> Over half of our engineers customize their programing font.  Among these users, [FiraCode](https://github.com/tonsky/FiraCode) was the most popular choice (those [programming ligatures](https://github.com/tonsky/FiraCode#whats-in-the-box) are amazing) followed by  [JetBrains Mono](https://www.jetbrains.com/lp/mono/). Other fonts include  [MonoStroom](https://github.com/Strum355/MonoStroom) , [Cascadia Code](https://github.com/microsoft/cascadia-code)),  and [Hack](https://sourcefoundry.org/hack/).

Reviewer: I had no idea that people did this. What is the benefit of customizing a programming font?

Reviewer 4: See the suggestion below.

## What we learned

> While these results are not statistically significant, one trend does stand out... developers will customize their environments to better suit their style of development. 

Reviewer: Why? I think we can pull this learning together with the intro do the workspaces. For example, 

"While these results are not statistically significant...suit their style of development. Personalizing your development environment does x,y,z. To show off just how different each developer's workspace can be, we pulled together X# of individual setups from Sourcegraphers. 

Move up the prompt and remove the header:
`What editors, themes and fonts do you use? DM or tweet us at [@srcgraph](https://twitter.com/srcgraph) to share your thoughts and images. /Make sure to not include sensitive information in any screens you share!/`

Add a header before the workspaces start: `## Workspaces of Sourcegraph`

Reviewer 4: I think the below captures this request?

---
title: The workspaces of Sourcegraph
author: Rob Rhyne
authorUrl: https://twitter.com/robrhyne
publishDate: 2021-02-10T18:00+02:00
tags: [blog, remote, design]
slug: workspaces-of-sourcegraph
published: true
---

> The Sourcegraph design team is exploring how to improve the display of code throughout the application. As part of this exploration, we are reviewing trends in typography, color schemes, and UI complexity in the places our engineers view code the most… their editors.  While our research is far from complete, we thought you might find a breakdown of the visual characteristics of where our team writes code of interest. 

Reviewer 4: ```suggestion
Developers read quite a bit. We read books, blog posts and documentation, but mostly, we read code. We read so much code, that many of us have become a bit particular about how it is displayed. 

So as the Sourcegraph design team began to study how to improve how we display code, it makes sense that we would study the place developers can decide exactly how it should look... their editors.

While our research is far from complete, we thought you might find a breakdown of the visual characteristics of where our team writes code of interest.
```

## Light/dark mode and contrast

> Three quarters of our engineers preferred a dark theme over a light theme.  Curiously, the majority of dark mode users chose a low contrast between code and background. Backgrounds colors averaged 85% black while the  primary code color to background contrast ratio was around 9 to 1. 

Reviewer 4: ```suggestion
Three-quarters of our engineers preferred a dark theme over a light theme.  Curiously, the majority of our dark mode users chose a lower contrast between code and background than their light theme oriented peers. Background colors averaged 85% black and the primary code color to background contrast ratio averaged 9 to 1.  This seems to contradict the dark themes offered by many other developer tools, where dark is often interpreted as near 100% black and text is often 100% white. 
```

Reviewer 4: @content-hacker I this will resolve your comment above?

Conversely, all light mode users at Sourcegraph prefer the maximum contrast with 100% white backgrounds and a black primary code color. 

![Representative contrast](https://sourcegraphstatic.com/blog/workspaces-of-sourcegraph/workspace-contrast.png)
*Contrast representative of typical dark and light modes.*

## Information density

> There are clearly two camps… those who use an IDE (mainly VS Code) and terminal based editors such as neovim/vim/emacs users. Editor users had around 96% of editor screen real estate dedicated to code. IDE users had from 80 to 90% of their screens is dedicated to code or the terminal, with the rest occupied by file trees and other UI such as menus and tabs. 

Reviewer 4: ```suggestion
There are clearly two camps… those who use an Integrated Developer Environment (mainly VS Code) and terminal-based editors such as neovim/vim/emacs users. Editor users had around 96% of editor screen real estate dedicated to code. IDE users had from 80 to 90% of their screens is dedicated to code or the terminal, with the rest occupied by file trees and other UI such as menus and tabs. 
```

## Fonts are commonly customized

> Over half of our engineers customize their programing font.  Among these users, [FiraCode](https://github.com/tonsky/FiraCode) was the most popular choice (those [programming ligatures](https://github.com/tonsky/FiraCode#whats-in-the-box) are amazing) followed by  [JetBrains Mono](https://www.jetbrains.com/lp/mono/). Other fonts include  [MonoStroom](https://github.com/Strum355/MonoStroom) , [Cascadia Code](https://github.com/microsoft/cascadia-code)),  and [Hack](https://sourcefoundry.org/hack/).

Reviewer 4: ```suggestion
Over half of our engineers customize their programing font.  Among these users, [FiraCode](https://github.com/tonsky/FiraCode) was the most popular choice, followed by  [JetBrains Mono](https://www.jetbrains.com/lp/mono/). Other fonts include  [MonoStroom](https://github.com/Strum355/MonoStroom) , [Cascadia Code](https://github.com/microsoft/cascadia-code)),  and [Hack](https://sourcefoundry.org/hack/).

Why customize your programming font? Glad you asked! Code, like no other language, can be dramatically altered by the slightest change. Programming fonts make it easier to distinguish characters and can render operators with [custom ligatures](https://github.com/tonsky/FiraCode#whats-in-the-box) to enhance readability. These traits can reduce errors and lead to better applications. 
```

## What we learned

> While these results are not statistically significant, one trend does stand out... developers will customize their environments to better suit their style of development. 

Reviewer 4: ```suggestion
While these results are not statistically significant, one trend does stand out... developers will customize their environments to improve the readability of code. Keep scrolling to check out some of the setups Sourcegraphers have deployed to help them create universal code search.

/What editors, themes and fonts do you use? DM or tweet us at @srcgraph to share your thoughts and images. Make sure to not include sensitive information in any screens you share!/

## Workspaces of Sourcegraph
```

- Mode: Dark
- Theme: SynthWave 84
- Extensions: Bracket Pair Colorizer, GitLens, Sourcegraph
- Misc: Fira Code font, Tilix terminal emulator

![Workspace of TJ Kandala, Software Engineer](https://storage.googleapis.com/sourcegraph-assets/blog/workspaces-of-sourcegraph/tj.png)

## Quinn Keast, Product Designer

- Nova (Cobalt theme) and VSCode (Cobalt2)
- Typeface: Operator Mono
- Sourcegraph light theme
- Desktop organization: Browser and reference materials on the left, Figma or full-screen editors on the main display. Sometimes I use terminals in the IDE, and other times I use iTerm2.

I'm experimenting with the Nova editor alongside VSCode, which replaced Sublime Text 2. I don't have any standard window setup, shifting as I need for any given design, research, or programming need. I've been using the fantastic Operator Mono font with ligatures and italics for years.

![Primary screen workspace organization of Quinn Keast, Product Designer](https://sourcegraphstatic.com/blog/workspaces-of-sourcegraph/Quinn-Keast-1.png)
![Secondary screen workspace organization of Quinn Keast, Product Designer](https://sourcegraphstatic.com/blog/workspaces-of-sourcegraph/Quinn-Keast-2.png)

> # What does your editor look like? 

Reviewer 4: ```suggestion
```

> What editors, themes and fonts do you use? DM or tweet us at [@srcgraph](https://twitter.com/srcgraph) to share your thoughts and images. /Make sure to not include sensitive information in any screens you share!/

Reviewer 4: ```suggestion
We want to see your setup too! Tweet it to us [@srcgraph](https://twitter.com/srcgraph).
```

Reviewer 4: @content-hacker Does it make sense to repeat the call for submissions? If so I changed the copy to not duplicate the one above.

---

> title: The workspaces of Sourcegraph

Reviewer: ```suggestion
title: Improving code display: A look at 17 developer workspaces 
```
